fix(metrics): Light normalize before extracting metrics [INGEST-1517 INGEST-1424]

iker-barriocanal opened the pull request:
# Context

Currently, all relay modes extract metrics from transactions. However, only processing relays are performing normalization operations that sample out events and modify the values that relay extracts metrics from, in the store normalization step. Not sampling out events early enough also introduces the consequence of extracting metrics from transactions that are discarded a bit later.

# What this PR does

Starting from `envelopes.rs`, there's a new function `light_normalize_event` to perform minimal event normalization to filter them later, and that minimal normalization has been extracted from the processors.

The changes in the processors are mostly refactors extracting functions to make them callable from the outside (from `light_normalize_event`, in this case). The normalization processor doesn't re-do the normalization work that was previously done in the light processing step.

An alternative to all this extraction into functions is to have another processor. This was discarded because of its complexity. No additional design choices have been considered.

# Other considerations

Performing light normalization in non-processing relays impacts any test depending on the amount of time relay takes to process an envelope. The cause is the long amount of time relay takes to initialize lazily the regex to parse user agents. Since that work is only done once, it doesn't impact relay instances on production.

iker-barriocanal committed:
extract security reports to light normalization

iker-barriocanal committed:
extract ip addr normalization to light normalization

iker-barriocanal committed:
move environment verification to light normalization

iker-barriocanal committed:
add _ to unused var

iker-barriocanal committed:
move release verification to light normalization

iker-barriocanal committed:
resort release <> environment check

iker-barriocanal committed:
tmp: update reqs comment

iker-barriocanal committed:
move release normalization to light normalization

iker-barriocanal committed:
move timestamp normalization to light normalization

iker-barriocanal committed:
move tag normalization to light normalization

iker-barriocanal committed:
move exception normalization to light normalization

iker-barriocanal committed:
move user agent normalization to light normalization

iker-barriocanal committed:
move measurement normalization to light normalization

iker-barriocanal committed:
move breakdown normalization to light normalization

iker-barriocanal committed:
extract transaction validation to fn from processor

iker-barriocanal committed:
validate transaction before light normalization

iker-barriocanal committed:
delete not needed comments

iker-barriocanal committed:
add temporary return types in main light normalization fn

iker-barriocanal committed:
return processing errors on light normalization

iker-barriocanal committed:
remove empty lines

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
+        normalize_exceptions(event)?;
+        normalize_user_agent(event, Some(true));
+        normalize_measurements(event);
+        normalize_breakdowns(event, breakdowns_config);
```
Would be nice to have a comment on each function call here, with the reason why it's needed for metrics extraction and/or inbound filters.

jjbayer replied:
I quickly checked these functions for idempotency, and I think they all are, but ideally we would add unit tests to verify that calling these functions twice does not change the result.

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
     }
 }
 
+pub fn light_normalize_event(
```
nit: We should probably come up with a better name for this function, but I don't have one yet. Maybe `pre_normalize_event`?

iker-barriocanal replied:
I hard agree "light normalization" isn't descriptive enough. I don't like `pre_normalize_*` because that indicates the operation happens _before_ normalization, and this function actually performs some normalization. I think that _some_ is what causes troubles for the naming. I'd also like to avoid referring to any filtering in the name, if possible -- the function may be used in another context in the future and shouldn't be tied up to current behavior.

What about something like `minimal_normalize_event`? That indicates there's some normalization happening, it doesn't perform the whole normalization, and the explanation of what it does and the motivation behind it could be added to the docstrings. What do you think?

jjbayer commented on relay-server/src/actors/envelopes.rs:
```diff
@@ -1853,6 +1853,36 @@ impl EnvelopeProcessor {
+            max_secs_in_future,
+            breakdowns_config,
+        )
+        .map_err(|e| ProcessingError::ProcessingFailed(e))?;
```
We could wrap this call in a `metric!(timer(...))` to measure how long light normalization takes.

jjbayer commented on relay-server/src/actors/envelopes.rs:
```diff
@@ -1882,6 +1912,8 @@ impl EnvelopeProcessor {
             self.finalize_event(state)?;
 
+            self.light_normalize_event(state)?;
+
```
With this, we can remove a lot of ad-hoc normalization from `extract_transaction_metrics`. But that can be a separate PR.

jan-auer replied:
Are we sure that double-normalizing (such as computing the breakdowns) doesn't break anything now?

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
+}
+
+/// Validates the timestamp range and sets a default value.
+fn normalize_timestamps(
```
A lot of these functions have been moved around, so it's hard to spot if any changes were made to them. Do you think we could leave them at their original position to reduce the diff?

If your intention was to make sure that functions are declared in the order that they are used, we can open a separate PR after this one that only reorders code.

jjbayer replied:
Ignore my previous comment, I just realized these used to be methods and now they are standalone functions.

jjbayer commented on relay-general/src/store/transactions.rs:
```diff
@@ -73,44 +116,7 @@ impl Processor for TransactionsProcessor {
-                ));
-            }
-        }
+        validate_transaction(event)?;
```
I think you mentioned it in the PR description, but couldn't we remove this call, now that this is already called from `light_normalize_event`?

iker-barriocanal replied:
That's a good question, and the motivation behind that decision isn't in the PR description. Normalizing the top-level transaction, in the current state of the PR, is indeed duplicated work. However, the processor also normalizes nested spans, and removing that validation step from the processor makes that function to not run on each span. I don't know and haven't investigated the consequences of doing that, but I don't want to take the risk there.

This explanation shouts for some refactoring to make the code easier to understand and not perform the duplicated work, but that is definitely out of scope for now.

jjbayer replied:
But the normalization of the spans happens on lines 121 - 131, right? The function `validate_transaction` does not touch spans as far as I can see, but I might be missing something.

Anyway I think we can leave this as-is, just to be extra sure, but I would add a code comment here that this is a candidate for refactoring.

jjbayer reviewed (commented):
This looks good overall! I did not yet verify whether everything that we need for inbound filters and / or metrics extraction is now actually part of `light_normalize_event`.

untitaker reviewed (commented):
please check out sentry, do `pip install -e ~/projects/relay/py/` and run its tests. I think you may break some tests by moving normalization out of the store normalizer, which we use in Python to fully normalize events.

It may be easier to introduce a flag on StoreNormalizer to opt into "light normalization", but it all depends on how many tests you're going to break.

iker-barriocanal committed:
add comments why each light normalization step is needed

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
-            }
-        }
+        #[cfg(not(feature = "uaparser"))]
+        panic!("relay not built with uaparser feature");
```
This looks risky, shall we just remove the `optional` in Cargo.toml and get rid of the feature gates?

https://github.com/getsentry/relay/blob/d91984ff4327973cb4f97b943736e834e1ee2b4d/relay-general/Cargo.toml#L38

iker-barriocanal committed:
add light normalization step to unit tests

iker-barriocanal committed:
add light normalization to the c abi

iker-barriocanal committed:
fix? cabi `relay_pii_strip_event`

jjbayer commented on relay-cabi/src/processing.rs:
```diff
@@ -151,6 +155,8 @@ pub unsafe extern "C" fn relay_pii_strip_event(
 
     let mut event = Annotated::<Event>::from_json((*event).as_str())?;
+    validate_annotated_transaction(&mut event)?;
+    light_normalize_event(&mut event, None, None, None, None, None, None)?;
```
Why add light normalization here? `processor` at this point is a `PiiProcessor`, not a `StoreProcessor`, so the call to `process_value` is not related to store normalization.

iker-barriocanal committed:
Revert "fix? cabi `relay_pii_strip_event`"

This reverts commit 25a9800dc0ee3436ca0235980658606fb7ba520e.

The processor called at this point is `PiiProcessor`, which doesn't need
light normalization.

iker-barriocanal committed:
extract all light normalization into a single function

before doing light normalization, it's required to validate a
transaction. this commit introduces a new function so that one doesn't
have to call the other two, and call this new function instead.

iker-barriocanal committed:
light normalize test_fixtures before store processing

warning: the snapshot update is the result of a breaking change
introduced in the light normalization, where the user agent is always
parsed no matter what the config looks like. previously, this was
disabled by default and you had to explicitly enable user agent
normalization. we accept this for now, and there will be work scheduled
in the future to address this.

jjbayer committed:
Merge remote-tracking branch 'origin/master' into iker/fix/filter-before-mep-extract

iker-barriocanal committed:
add _ in unused var

jjbayer committed:
fix: Do not emit transaction.op=default as metric tag value

iker-barriocanal committed:
delete `uaparser` cfg

its risky to leave the `uaparser` there, potentially panicking. removing
the cfg makes it required though.

iker-barriocanal committed:
update tests

due to the static regex to parse user agents being slow in
initialization, integration tests need much more time to run. as a
result, the timeout has been increased from 1s to ~7s, depending on the
test.

iker-barriocanal committed:
log metric for light normalization

iker-barriocanal committed:
update changelog

iker-barriocanal committed:
move test import to test section

also makes compiler happy

iker-barriocanal committed:
fix linting

iker-barriocanal commented on tests/integration/test_security_report.py:
```diff
@@ -126,14 +126,19 @@ def test_security_report(mini_sentry, relay, test_case, json_fixture_provider):
     ext = ".no_processing.output"
     expected_evt = fixture_provider.load(test_name, ext)
 
+    if "received" in event:
+        event.pop("received")
+    if "timestamp" in event:
+        event.pop("timestamp")
+
```
At least in the `expect_ct` case the event has the `received` and `timestamp` fields. I haven't thought if we want them, but I don't think this is that important right now and I took the short and hacky path deleting these fields here. We may want to revisit this.

jjbayer replied:
I think this is fine, just a result of timestamp normalization.

iker-barriocanal reviewed (commented):
There's a decent amount of stuff added to the output of integration tests of security reports. By looking at the code, all changes seem they make sense to me, so I assume they weren't running before.

jjbayer commented on tests/integration/test_security_report.py:
```diff
@@ -46,7 +46,7 @@ def test_uses_origins(mini_sentry, relay, json_fixture_provider, allowed_origins
 
     if should_be_allowed:
-        mini_sentry.captured_events.get(timeout=1).get_event()
+        mini_sentry.captured_events.get(timeout=10).get_event()
```
Note: This is necessary because the non-processing relay now does normalization, which includes UA parsing, which apparently has slow initialization.

jjbayer commented on CHANGELOG.md:
```diff
@@ -15,7 +15,7 @@
 - Change to the internals of the healthcheck endpoint. ([#1374](https://github.com/getsentry/relay/pull/1374))
 - Re-encode the Typescript payload to normalize. ([#1372](https://github.com/getsentry/relay/pull/1372))
-
+- Light normalize before extracting metrics. ([#1366](https://github.com/getsentry/relay/pull/1366))
```
```suggestion
- Partially normalize events before extracting metrics. ([#1366](https://github.com/getsentry/relay/pull/1366))
```

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -907,6 +939,7 @@ fn test_user_ip_from_remote_addr() {
 
     let config = StoreConfig::default();
     let mut processor = NormalizeProcessor::new(Arc::new(config), None);
+    light_normalize(&mut event, None, None, None, None, None, None).unwrap();
```
nit: We should really pass a config struct instead of passing this many arguments, but let's leave that refactor for a follow-up PR.

jjbayer commented on relay-general/src/store/mod.rs:
```diff
@@ -137,3 +140,25 @@ impl<'a> Processor for StoreProcessor<'a> {
+    breakdowns_config: Option<BreakdownsConfig>,
+) -> ProcessingResult {
+    validate_annotated_transaction(event)?;
+    light_normalize_event(
```
nit: If my search is correct, `light_normalize_event` is never called outside of `light_normalize`. So why not simplify this by calling `validate_annotated_transaction` from inside `light_normalize_event` (and then remove the `light_normalize` wrapper)?

iker-barriocanal replied:
Good feedback, I'll do that in a follow-up PR.

jan-auer replied:
+1 on this, would even prefer if this goes into the same PR.

iker-barriocanal replied:
Addressed in https://github.com/getsentry/relay/pull/1385.

iker-barriocanal committed:
improve changelog message

jan-auer commented on relay-general/src/store/mod.rs:
```diff
@@ -8,7 +8,9 @@ use serde_json::Value;
-use crate::types::{Meta, ProcessingResult, SpanAttribute};
+use crate::types::{Annotated, Meta, ProcessingResult, SpanAttribute};
+use normalize::light_normalize_event;
+use transactions::validate_annotated_transaction;
```
(before merging) Please qualify functions at use site.

jan-auer commented on relay-general/src/store/mod.rs:
```diff
@@ -137,3 +140,25 @@ impl<'a> Processor for StoreProcessor<'a> {
+    received_at: Option<DateTime<Utc>>,
+    max_secs_in_past: Option<i64>,
+    max_secs_in_future: Option<i64>,
+    breakdowns_config: Option<BreakdownsConfig>,
```
(before merging) These are now enough parameters to warrant some form of config struct like we have in the store normalizer.

jan-auer commented on relay-general/src/store/normalize.rs:
```diff
@@ -344,166 +211,355 @@ impl<'a> NormalizeProcessor<'a> {
+        })?;
+
+        // Default required attributes, even if they have errors
+        normalize_release_dist(event); // dist is a tag extracted along with other metrics from transactions
```
_nit_: Do we need all these comments here?

jjbayer replied:
I asked for these comments because I figured it might be helpful to know in the future why we decided to pull those into "light" normalization.

jan-auer reviewed (commented):
I believe, we need to run the `SchemaProcessor` at the very least since that enforces things like: `max_chars`, `required`, `trim_whitespace`, and `nonempty`. In store normalization, the schema processor runs before the normalizer, which means that it can change values in such a way that the normalizer behaves differently.

iker-barriocanal committed:
qualify some functions at use site

iker-barriocanal committed:
move schema processor to store normalization

iker-barriocanal committed:
Move light normalization params into config struct

iker-barriocanal committed:
add light normalization idempotency test

iker-barriocanal committed:
fix lint

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -823,6 +830,18 @@ impl Default for NormalizeProcessor<'_> {
+        max_secs_in_future: None,
+        breakdowns_config: None,
+    }
+}
```
nit: You can put `#[derive(default)]` above `struct LightNormalizationConfig`, and then call `Default::default()` to get an empty config.

iker-barriocanal replied:
Addressed in https://github.com/getsentry/relay/pull/1386.

jjbayer reviewed (approved):
Feel free to ignore nitpick.
